discovery/azure/azure.go

```diff
@@ -413,26 +413,20 @@ func (d *Discovery) refresh(ctx context.Context) ([]*targetgroup.Group, error) {
 					} else {
 						networkInterface, err = client.getVMScaleSetVMNetworkInterfaceByID(ctx, nicID, vm.ScaleSet, vm.InstanceID)
 					}
+
 					if err != nil {
 						if errors.Is(err, errorNotFound) {
 							level.Warn(d.logger).Log("msg", "Network interface does not exist", "name", nicID, "err", err)
 						} else {
 							ch <- target{labelSet: nil, err: err}
 						}
-						d.addToCache(nicID, networkInterface)
-					} else {
-						networkInterface, err = client.getVMScaleSetVMNetworkInterfaceByID(ctx, nicID, vm.ScaleSet, vm.InstanceID)
```

Comment:
It's worth checking why we ended up with "duplicated" code like this, maybe it's a bad merge. Maybe that merge broke other things...

Normally, https://github.com/prometheus/prometheus/commit/0704c7254876fefbdde85fecb8ad389106312af2 already brought the code to the state you requested.

Also it'd be great if we have a regression test for this.

Reply:
Good catch by the way!

Reply:
> It's worth checking why we ended up with "duplicated" code like this, [...]

I'm not sure if we already have a tool to check for "duplicate code." If not, I recommend that we consider using a code quality scanner tool https://www.sonarsource.com/blog/manage-duplicated-code-with-sonar/ (open source code available at https://github.com/SonarSource/sonarqube).

> Normally, https://github.com/prometheus/prometheus/commit/0704c7254876fefbdde85fecb8ad389106312af2 already brought the code to the state you requested.

Do you want to revert that commit, or should I leave the pull request open until it gets merged? I encountered this issue actually 😅

> Also it'd be great if we have a regression test for this.

I think we could create a new issue for that and see if others can help with it.

Reply:
I was trying to understand how we ended up with the code in main because https://github.com/prometheus/prometheus/commit/0704c7254876fefbdde85fecb8ad389106312af2 (that was already merged into main) already brought the code to the state of your PR.

I think the merge https://github.com/prometheus/prometheus/pull/13399 is the one that added that duplicated/stale code back by mistake.

I'll let the maintainers decide for the need of a regression test.